waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/PanopticOperationHandler.java

```diff
@@ -15,30 +15,82 @@
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiFunction;
 
 import org.apache.hadoop.hive.metastore.HiveMetaStore.HMSHandler;
 import org.apache.hadoop.hive.metastore.api.TableMeta;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.hotels.bdp.waggledance.mapping.model.DatabaseMapping;
+import com.hotels.bdp.waggledance.mapping.service.requests.GetAllDatabasesByPatternRequest;
+import com.hotels.bdp.waggledance.mapping.service.requests.GetAllDatabasesRequest;
+import com.hotels.bdp.waggledance.mapping.service.requests.GetTableMetaRequest;
+import com.hotels.bdp.waggledance.mapping.service.requests.SetUgiRequest;
 
 /**
  * Class responsible for handling the Hive operations that need to combine results from multiple Hive Metastores
  */
-public interface PanopticOperationHandler {
+public abstract class PanopticOperationHandler {
+
+  protected static final long GET_DATABASES_TIMEOUT = TimeUnit.MILLISECONDS.toMillis(8000L);
+  private static final Logger LOG = LoggerFactory.getLogger(PanopticOperationHandler.class);
+  private static final String INTERRUPTED_MESSAGE = "Execution was interrupted: ";
+  private static final String SLOW_METASTORE_MESSAGE = "Metastore {} was slow to respond so results are omitted";
+  private static final long SET_UGI_TIMEOUT = TimeUnit.MILLISECONDS.toMillis(8000L);
+  private static final long GET_TABLE_META_TIMEOUT = TimeUnit.MILLISECONDS.toMillis(400L);
 
   /**
    * Implements {@link HMSHandler#get_all_databases()} over multiple metastores
    *
    * @return list of all databases
    */
-  List<String> getAllDatabases();
+  public abstract List<String> getAllDatabases();
 
   /**
    * Implements {@link HMSHandler#get_database(String)} over multiple metastores
    *
    * @param databasePattern pattern to match
    * @return list of all databases that match the passed pattern
    */
-  List<String> getAllDatabases(String databasePattern);
+  public abstract List<String> getAllDatabases(String databasePattern);
+
+  protected List<String> getAllDatabases(
+      Map<DatabaseMapping, String> databaseMappingsForPattern,
+      BiFunction<String, DatabaseMapping, Boolean> filter) {
+    ExecutorService executorService = Executors.newFixedThreadPool(databaseMappingsForPattern.size());
+    List<GetAllDatabasesByPatternRequest> allRequests = new ArrayList<>();
+    long maxLatency = (long) Integer.MIN_VALUE;
+
+    for (Entry<DatabaseMapping, String> mappingWithPattern : databaseMappingsForPattern.entrySet()) {
+      DatabaseMapping mapping = mappingWithPattern.getKey();
+      GetAllDatabasesByPatternRequest databasesByPatternRequest = new GetAllDatabasesByPatternRequest(
+          mapping, mappingWithPattern.getValue(), filter);
+      allRequests.add(databasesByPatternRequest);
+      maxLatency = Math.max(maxLatency, mapping.getLatency());
+    }
+
+    long totalTimeout = Math.max(1, GET_DATABASES_TIMEOUT + maxLatency);
```

Comment:
perhaps a method: `long getTotalTimeout(GET_DATABASES_TIMEOUT , databaseMappingsForPattern.keySet())` and use that. So the maxLatency and Math.max(1...) stuff is not copied in all the places?

Reply:
Done